### PR TITLE
Create Elasticsearch River with last sequence param

### PR DESCRIPTION
**Pull request by markbreedlove**
- Change River creation to require a last sequence number parameter, and to set it on the newly-created River.
- Leave `V1::SearchEngine` shims for methods in `V1::SearchEngine::River` in-place but modified them where necessary, related to the River creation change.
- Update River re-creation to get the last sequence number of the old River and set it for the new one.
- Alter the behavior of `V1::SearchEngine::River.last_sequence` for it to raise an exception if there is an HTTP error, such as will happen if the river does not exists, or exists, but it pointed at an empty CouchDB database without having a `last_seq` number set. This should be an edge case now that we're having create_river require a `last_seq` parameter.
- Add a number of missing unit tests for `V1::SearchEngine::River`.
- Add status messages to stdout.
- Add doc comments.

Rake task changes:
- `create_river`: Task description no longer suggests defaulting to the currently-deployed index (implies that you should specify it). I'm not sure anyway that you could have left it empty in the rake task arguments.
- `delete_river`: Now requires that you enter the name of the river to delete.

Addresses https://issues.dp.la/issues/8576

### CODE REVIEW

**Comment by markbreedlove** [comment]
I've got at least one thing to fix, which was turned up by Travis.

@mdellabitta Sounds like a good idea to have the verification. I'm looking again more closely to see how it's called.

**Comment by markbreedlove** [comment]
@mdellabitta I've gone back and had a second look at `recreate_river`. The only time that it would do any harm and overwrite ES index documents would be if the old `last_seq` was already "0". The only place it's called outside of the `recreate_river` rake task is in `V1::Repository.recreate_env`, which is called from the `recreate_repo_env` rake task, which has always been buggy. `recreate_repo_env` is a development task for wiping clean your development instance and starting over, so it actually needs to be fixed to be _more_ destructive and call `V1::SearchEngine::River.delete_river` and `create_river("0")`.

I'll make some tickets later for changing `recreate_repo_env` and actually adding a rake task for the lonely old `recreate_repo_env_with_docs` method (only used by a cucumber test), which could be useful now that we're having a phase of renewed development with the legacy system.

**Comment by markbreedlove** [comment]
Reviewers please note that I forgot to fill in the spec test for `V1::SearchEngine::River.recreate_river`! I'll do that soon.

File: v1/lib/v1/search_engine/river.rb [hunk 1]
```diff
@@ -165,29 +187,75 @@ def self.verify_river_status(name=river_name)
...
-
-        return nil if seq.nil?
+        response = HTTParty.get(endpoint(name) + '/_seq')
+        raise "Problem getting last sequence: #{JSON.parse(response.body)}" \
+          unless response.success?
+        seq = response.parsed_response['_source']['couchdb']['last_seq']
 
         # handle bigcouch's last_seq format if it looks like it
-        seq = JSON.parse(seq).first if seq !~ /^\d+$/
+        seq = JSON.parse(seq).first if seq !~ /^\d+$/  # but how not?  -mb
```
Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: Returning an `Integer` here is probably fine for human legibility and the sake of comparison, but review my comment on `#last_sequence!`

File: v1/lib/v1/search_engine/river.rb [hunk 2]
```diff
@@ -165,29 +187,75 @@ def self.verify_river_status(name=river_name)
...
 
+      ##
+      # Set the last sequence number for a River.
+      #
+      # The sequence number can be PUT for a River before the rest of it is
+      # created by a PUT to its _meta resource. Putting last_seq creates a data
+      # structure for the incomplete river. When you do this (before PUTing to
+      # `_meta') you'll see a property for the new river in `/_river/_mapping'.
+      #
+      # @param  seq_num [Fixnum or String] Sequence number /^\d+$/
```
Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: So, BigCouch reports a "sequence number" as a JSON array; first item is equivalent to the CouchDB last_seq, and the second is a long, opaque string. Upon river creation, the full array format is necessary, otherwise the call to BigCouch will fail: 

```
$ curl -sv "repo-stg1:5984/dpla/_changes?since=1"
* Connected to repo-stg1 port 5984 (#0)
> GET /dpla/_changes?since=1 HTTP/1.1
> Host: repo-stg1:5984
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 500 Internal Server Error
< X-Couch-Request-ID: 3254ddde
< Server: CouchDB/1.1.1 (Erlang OTP/R14B01)
< Date: Thu, 29 Sep 2016 16:14:20 GMT
< Content-Type: text/plain;charset=utf-8
< Content-Length: 164
< Cache-Control: must-revalidate
<
{"error":"function_clause","reason":null,"stack":["chttpd_db:changes_callback/2","chttpd:handle_request/1","mochiweb_http:headers/5","proc_lib:init_p_do_apply/3"]}
```

File: v1/lib/v1/search_engine/river.rb [hunk 3]
```diff
@@ -165,29 +187,75 @@ def self.verify_river_status(name=river_name)
...
+      # @return true
+      # @raise [RuntimeError]  if seq_num is improper type
+      # @raise [RuntimeError]  if HTTP transaction fails
+      #
+      def self.last_sequence!(seq_num, rname=river_name)
+        # rname will be present but nil when called from a rake task
+        rname ||= river_name
+        seq_num_s = seq_num.to_s
+        raise "Bad sequence number #{seq_num_s}" if seq_num_s !~ /^\d+$/
```
Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: I think we shouldn't do this check because of my comment above.

File: v1/lib/v1/search_engine/river.rb (hunk 2 above)
Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: Also, just for reference, the river stores the `last_seq` value as a string, regardless of the incoming format. A quick glance at the river code suggests that this won't cause issues for us, but I haven't tested it.

File: v1/lib/v1/search_engine/river.rb (hunk 3 above)
Comment (markbreedlove) [inline v1/lib/v1/search_engine/river.rb]: I know about the array format, but I'm not sure I understand, because the only way `last_sequence!` is going to get `seq_num` is from the call in `.create_river`, where the value is either going to be an integer (which a calling function has got from `.last_sequence` ) or a string that's been entered by a person invoking the rake task (where the person should be in control of putting in a number). Do you know of another place where we might be getting some non-numeric value? Note that we're getting the last sequence number in our code from the ES river endpoint (`/_river/rivername/_seq`) and that parses out and returns the numeric value.  Example:
```
{
  "_index": "_river",
  "_type": "mbtest_river",
  "_id": "_seq",
  "_version": 1,
  "exists": true,
  "_source": {
    "couchdb": {
      "last_seq": "0"
    }
  }
}
```

File: v1/lib/v1/search_engine/river.rb (hunk 2 above)
Comment (markbreedlove) [inline v1/lib/v1/search_engine/river.rb]: The ES River stores it as an integer if you PUT an integer, string if you give it a string. For example:
```
{
  "_index": "_river",
  "_type": "mbtest_river",
  "_id": "_seq",
  "_version": 2,
  "exists": true,
  "_source": {
    "couchdb": {
      "last_seq": 1
    }
  }
}
```
... after PUTing integer 1.

The cast to integer is something that was already there. I'd have to do more of an audit to see how necessary it is.

File: v1/lib/v1/search_engine/river.rb (hunk 3 above)
Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: The regex here is matching digits only. The code in the the ES river plugin just takes whatever it's given and passes it along to a request as the value for the `since` parameter to the [CouchDB `_changes` endpoint](http://docs.couchdb.org/en/1.6.1/api/database/changes.html).  As indicated [above](https://github.com/dpla/platform/pull/42/files#r81175160) the request that the river ends up sending to Bigcouch will fail. Here's what our prod river endpoint is returning:

```
deploy mark$ curl -s  "http://search-prod1:9200/_river/dpla_river/_seq" && echo
{"_index":"_river","_type":"dpla_river","_id":"_seq","_version":5928619,"exists":true, "_source" : {"couchdb":{"last_seq":"[91433533,\"g1AAAAF7eJzLYWBg4MhgTmEQTspMT84vTc5wKEotyNctKMpPMcwBSjIlMiTJ____PyuJYV36FRwqjcAqkxSAZJI9VHHlTPyKHUCK46GKJ-rgV5wAUlwPVRz6B6_iPBYgydAApIDq54M0hD0jQsMCiIb9IA3-L4nQcACi4T5IQ8FsvGEI0fAAogHsBy-7LAA8doB-\"]"}}}
```

File: v1/lib/v1/search_engine/river.rb (hunk 2 above)
Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: Sorry - to clarify, the *indexer* code in the river updates the `.../_seq` and stores the value as a string.

File: v1/lib/v1/search_engine/river.rb (hunk 3 above)
Comment (markbreedlove) [inline v1/lib/v1/search_engine/river.rb]: That's really interesting. Now I see what you mean. I had an old Postman tab open from yesterday, for my local VM (Bigcouch, same version as staging & production). I had a result for `loadbal:9201/_river/dpla_river/_seq` that contained a `last_seq` value of a long array similar to what you're seeing.  I reloaded the GET request today and now it is a numeric string: "14119".

Elsewhere, I see the same array you're reporting on staging. However, in developing this, I am unable to reproduce the array result, reloading `_seq`.

Also note my comment below, which I accidentally posted outside of this thread, about how it can be integer or string.

The issue remains, though, for me, that this method, `last_sequence!`, which is for setting the sequence number, should be able to expect a numeric argument. Whatever the `GET .../_seq` endpoint returns, it's parsed by `last_sequence` or by the user who types it.

Comment (anarchivist) [inline v1/lib/v1/search_engine/river.rb]: > The issue remains, though, for me, that this method, `last_sequence!`, which is for setting the sequence number, should be able to expect a numeric argument. Whatever the `GET .../_seq` endpoint returns, it's parsed by `last_sequence` or by the user who types it.

Agreed, but if I'm not mistaken here the regex is still a problem, because it will `raise` when `seq_num_s` doesn't match, e.g. when it runs into anything that is not a set of 1 or more digits - like the array sequence numbers.

Comment (markbreedlove) [inline v1/lib/v1/search_engine/river.rb]: Right, but `last_sequence` can't return that array or JSON string, it can only return an integer; and the user running `bundle exec rake v1:create_river[index, rivername, 123]` can be expected to enter a numeric string. Otherwise, I can't determine how this function would receive a `seq_num` argument that is not numeric. I can only see that happening if someone typed it as the third argument to the rake task, in which case I think it's fair to throw an exception.

**Comment by markbreedlove** [comment]
https://github.com/dpla/platform/pull/42/commits/6ccef482fd07f1cbd542f7f8666a23d9509e0987 will make index creation happen with the big JSON array string if that's what comes back as the last sequence value. The integer casting has been broken out to its own function that is only called where it's needed.

**Comment by anarchivist** [comment]
I'm 👍 - others want to review the rest of the changes?

**Comment by moltude** [comment]
I think this is 👍  too
